Assets/Mapzen/PolylineBuilder.cs

```diff
@@ -88,11 +122,110 @@ public void OnEndPolygon()
         {
         }
 
-        public static Vector2 Perp(Vector2 d)
+        private static Vector2 Perp(Vector2 d)
         {
             var p = new Vector2(-d.y, d.x);
             return p.normalized;
         }
+
+        private static bool Intersection(Vector2 p0, Vector2 p1, Vector2 v0, Vector2 v1, out Vector2 intersection)
+        {
+            intersection = new Vector2();
+
+            var p0p1 = p1 - p0;
+
+            var p0p1crossv0 = p0p1.x * v0.y - p0p1.y * v0.x;
+            var p0p1crossv1 = p0p1.x * v1.y - p0p1.y * v1.x;
+
+            var v1crossv0 = (v1.x * v0.y - v1.y * v0.x);
+
+            float u = p0p1crossv1 / v1crossv0;
+            float t = p0p1crossv0 / v1crossv0;
+
+            // Collinear or parallel
+            if (Math.Abs(v1crossv0) <= float.Epsilon)
+            {
+                return false;
+            }
+
+            if (u >= 0.0f && u <= 1.0f && t >= 0.0f && t <= 1.0f)
+            {
+                intersection = p0 + t * v1;
+                return true;
+            }
+
+            return false;
+        }
+
+        private void AddPoint(bool isFirstPoint, Vector2 lastPoint, Vector2 currPoint, Vector2 nextPoint)
+        {
+            float extrude = options.Width * 0.5f;
+
+            if (isFirstPoint)
+            {
+                Vector2 perp = currPoint == lastPoint ?
+                    Perp(nextPoint - currPoint) * extrude :
+                    Perp(currPoint - lastPoint) * extrude;
+
+                polygonBuilder.OnPoint(new Point(lastPoint.x - perp.x, lastPoint.y - perp.y));
+                polygonBuilder.OnPoint(new Point(lastPoint.x + perp.x, lastPoint.y + perp.y));
+            }
+
+            if (currPoint == lastPoint || nextPoint == currPoint)
+            {
+                return;
+            }
+
+            var n0 = Perp(currPoint - lastPoint) * extrude;
+            var n1 = Perp(nextPoint - currPoint) * extrude;
+
+            bool isRightTurn = n1.x * n0.y - n1.y * n0.x > 0.0f;
+
+            if (isRightTurn)
```

Comment:
This section could benefit from comments explaining how the miter limit is applied.